collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py

```diff
@@ -158,66 +155,127 @@ def get_document(filepath, line):
+        a_document = get_document(
+            a_filename, self.sentence_indices[a_filename][a_line_idx])
+        a_document, a_line_idx = match_target_seq_length(
+            a_document, target_seq_length_a, a_filename, a_line_idx,
+            self.sentence_indices)
 
-            b_filename = a_filename
-            b_line = a_line
-            b_document = a_document
-        else:
-            # The rest of the time, B is a random sentence from the corpus
+        if self.sentence_indices[a_filename][a_line_idx] >= \
+                self.sentence_indices[a_filename][-1] or \
+                random.random() < 0.5:
+
+            # About 50% of the time, B is a random sentence from the corpus
             label = 0
 
-            a_start_idx = random.randrange(len(a_document) - a_length)
+            for _ in range(10):
+                b_filename = random.choice(self.filenames)
+                b_line_idx = random.choice(
+                    range(len(self.sentence_indices[b_filename])))
+                if b_filename != a_filename:
+                    break
+                else:
+                    new_idx = self.sentence_indices[b_filename][b_line_idx]
+                    old_idx = self.sentence_indices[b_filename][a_line_idx]
+                    if (new_idx < old_idx - self.max_seq_length) or \
+                            (new_idx > old_idx + self.max_seq_length):
+                        break
+                    else:
+                        pass
+
+            b_document = get_document(
+                b_filename, self.sentence_indices[b_filename][b_line_idx])
+            b_document, b_line_idx = match_target_seq_length(
+                b_document, target_seq_length_b, b_filename, b_line_idx,
+                self.sentence_indices)
+
+        else:
+            label = 1
 
-            b_filename = random.choice(self.filenames)
-            b_line = random.choice(self.sentence_indices[b_filename])
-            b_document = get_document(b_filename, b_line)
-            b_start_idx = random.randrange(len(b_document) - b_length)
+            b_filename = a_filename
+            b_line_idx = a_line_idx + 1
+            b_document = get_document(
+                a_filename, self.sentence_indices[a_filename][b_line_idx])
+            b_document, b_line_idx = match_target_seq_length(
+                b_document, target_seq_length_b, b_filename, b_line_idx,
+                self.sentence_indices)
+
+        def truncate_seq_pair(a_document, b_document, max_num_tokens):
+            # Truncates a pair of sequences to a maximum sequence length
+            while True:
+                total_length = len(a_document) + len(b_document)
+                if total_length <= max_num_tokens:
+                    break
+
+                trunc_document = a_document if len(
+                    a_document) > len(b_document) else b_document
+                assert len(trunc_document) >= 1
+
+                if random.random() < 0.5:
```

Comment:
I don't understand what this part is doing, could you please elaborate. This 0.5 isn't related to the 0.5 above, right?